File: app/src/main/java/com/attilapalfi/exceptional/ui/main/exception_instance_views/ExceptionInstancesFragment.java
package com.attilapalfi.exceptional.ui.main.exception_instance_views;

import java.math.BigInteger;
import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.format.DateFormat;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;
import com.attilapalfi.exceptional.R;
import com.attilapalfi.exceptional.dependency_injection.Injector;
import com.attilapalfi.exceptional.interfaces.ExceptionChangeListener;
import com.attilapalfi.exceptional.interfaces.ExceptionRefreshListener;
import com.attilapalfi.exceptional.model.Exception;
import com.attilapalfi.exceptional.model.Friend;
import com.attilapalfi.exceptional.persistence.ExceptionInstanceStore;
import com.attilapalfi.exceptional.persistence.FriendStore;
import com.attilapalfi.exceptional.persistence.ImageCache;
import com.attilapalfi.exceptional.persistence.MetadataStore;
import com.attilapalfi.exceptional.rest.ExceptionService;
import com.attilapalfi.exceptional.ui.main.friends_page.FriendDetailsActivity;

/**
 * Created by palfi on 2015-08-24.
 */
public class ExceptionInstancesFragment extends Fragment implements ExceptionRefreshListener,
        ExceptionChangeListener, SwipeRefreshLayout.OnRefreshListener {

    private static long lastSyncTime = 0;

    @Inject
    ExceptionService exceptionService;
    @Inject
    ExceptionInstanceStore exceptionInstanceStore;
    @Inject
    MetadataStore metadataStore;
    private Friend friend;
    private RecyclerView recyclerView;
    private ExceptionInstanceAdapter exceptionInstanceAdapter;
    private SwipeRefreshLayout swipeRefreshLayout;

    @Override
    public void onCreate( @Nullable Bundle savedInstanceState ) {
        super.onCreate( savedInstanceState );
        Injector.INSTANCE.getApplicationComponent().inject( this );
        exceptionInstanceStore.addExceptionChangeListener( this );
        FriendDetailsActivity activity;
        if ( getActivity() instanceof FriendDetailsActivity ) {
            activity = (FriendDetailsActivity) getActivity();
            friend = activity.getFriend();
        }
    }

    @Nullable
    @Override
    public View onCreateView( LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState ) {
        initExceptionAdapter();
        View view = initRecyclerView( inflater, container );

        swipeRefreshLayout = (SwipeRefreshLayout) view.findViewById( R.id.exception_swipe_container );
        swipeRefreshLayout.setOnRefreshListener( this );
        swipeRefreshLayout.setSize( SwipeRefreshLayout.LARGE );
        swipeRefreshLayout.setColorSchemeColors(
                getResources().getColor( R.color.exceptional_blue ),
                getResources().getColor( R.color.exceptional_green ),
                getResources().getColor( R.color.exceptional_red ),
                getResources().getColor( R.color.exceptional_purple )
        );

        return view;
    }

    @Override
    public void onActivityCreated( @Nullable Bundle savedInstanceState ) {
        super.onActivityCreated( savedInstanceState );
        onExceptionsChanged();
    }

    @Override
    public void onDestroy( ) {
        exceptionInstanceStore.removeExceptionChangeListener( this );
        super.onDestroy();
    }

    @Override
    public void onExceptionsChanged( ) {
        exceptionInstanceAdapter.notifyDataSetChanged();
    }

    @Override
    public void onRefresh( ) {
        long currentTime = System.currentTimeMillis();
        if ( currentTime > lastSyncTime + 10000 ) {
            lastSyncTime = currentTime;
        }
        actualRefresh();
    }

    private void actualRefresh( ) {
        if ( metadataStore.isLoggedIn() ) {
            exceptionService.refreshExceptions( this );
        } else {
            Toast.makeText( getActivity().getApplicationContext(), "You have to login first!", Toast.LENGTH_SHORT ).show();
            onExceptionRefreshFinished();
        }
    }

    @Override
    public void onExceptionRefreshFinished( ) {
        swipeRefreshLayout.setRefreshing( false );
    }

    private void initExceptionAdapter( ) {
        List<Exception> values = generateValues();
        exceptionInstanceAdapter = new ExceptionInstanceAdapter( values, getActivity().getApplicationContext() );
        onExceptionsChanged();
    }

    private List<Exception> generateValues( ) {
        List<Exception> allValues = exceptionInstanceStore.getExceptionList();
        if ( friend != null ) {
            List<Exception> filteredValues = new ArrayList<>();
            for ( Exception exception : allValues ) {
                if ( exception.getFromWho().equals( friend.getId() ) || exception.getToWho().equals( friend.getId() ) ) {
                    filteredValues.add( exception );
                }
            }
            return filteredValues;
        } else {
            return allValues;
        }
    }

    @NonNull
    private View initRecyclerView( LayoutInflater inflater, ViewGroup container ) {
        View fragmentView = inflater.inflate( R.layout.fragment_exception_instances, container, false );
        recyclerView = (RecyclerView) fragmentView.findViewById( R.id.exception_recycler_view );
        recyclerView.setLayoutManager( new LinearLayoutManager( getActivity() ) );
        recyclerView.setAdapter( exceptionInstanceAdapter );
        exceptionInstanceAdapter.setRecyclerView( recyclerView );
        return fragmentView;
    }
}

File: app/src/main/java/com/attilapalfi/exceptional/ui/main/exception_instance_views/ExceptionInstancesFragment.kt
package com.attilapalfi.exceptional.ui.main.exception_instance_views

import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.content.ContextCompat
import android.support.v4.widget.SwipeRefreshLayout
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import com.attilapalfi.exceptional.R
import com.attilapalfi.exceptional.dependency_injection.Injector
import com.attilapalfi.exceptional.interfaces.ExceptionChangeListener
import com.attilapalfi.exceptional.interfaces.ExceptionRefreshListener
import com.attilapalfi.exceptional.model.Exception
import com.attilapalfi.exceptional.model.Friend
import com.attilapalfi.exceptional.persistence.ExceptionInstanceStore
import com.attilapalfi.exceptional.persistence.MetadataStore
import com.attilapalfi.exceptional.rest.ExceptionRestConnector
import com.attilapalfi.exceptional.ui.main.friends_page.FriendDetailsActivity
import jp.wasabeef.recyclerview.animators.adapters.ScaleInAnimationAdapter
import java.util.*
import javax.inject.Inject

/**
 * Created by palfi on 2015-08-24.
 */
class ExceptionInstancesFragment : Fragment(), ExceptionRefreshListener, ExceptionChangeListener,
        SwipeRefreshLayout.OnRefreshListener {

    @Inject
    lateinit var exceptionRestConnector: ExceptionRestConnector
    @Inject
    lateinit var exceptionInstanceStore: ExceptionInstanceStore
    @Inject
    lateinit var metadataStore: MetadataStore
    private var friend: Friend? = null
    private lateinit var recyclerView: RecyclerView
    private lateinit var adapter: ScaleInAnimationAdapter
    private lateinit var swipeRefreshLayout: SwipeRefreshLayout
    private var refreshing = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        Injector.INSTANCE.applicationComponent.inject(this)
        exceptionInstanceStore.addExceptionChangeListener(this)
        val activity: FriendDetailsActivity
        if (getActivity() is FriendDetailsActivity) {
            activity = getActivity() as FriendDetailsActivity
            friend = activity.friend
        }
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = initRecyclerView(inflater, container)
        initSwipeLayout(view)
        return view
    }

    private fun initSwipeLayout(view: View) {
        swipeRefreshLayout = view.findViewById(R.id.exception_swipe_container) as SwipeRefreshLayout
        swipeRefreshLayout.setOnRefreshListener(this)
        swipeRefreshLayout.setSize(SwipeRefreshLayout.LARGE)
        swipeRefreshLayout.setColorSchemeColors(
                ContextCompat.getColor(context, R.color.exceptional_blue),
                ContextCompat.getColor(context, R.color.exceptional_green),
                ContextCompat.getColor(context, R.color.exceptional_red),
                ContextCompat.getColor(context, R.color.exceptional_purple)
        )
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        onExceptionsChanged()
    }

    override fun onDestroy() {
        exceptionInstanceStore.removeExceptionChangeListener(this)
        super.onDestroy()
    }

    override fun onExceptionsChanged() {
        (adapter.wrappedAdapter as ExceptionInstanceAdapter).setValues(generateValues())
        (adapter.wrappedAdapter as ExceptionInstanceAdapter).notifyDataSetChanged()
    }

    override fun onRefresh() {
        if (!refreshing) {
            refreshing = true
            val currentTime = System.currentTimeMillis()
            if (currentTime > lastSyncTime + 10000) {
                lastSyncTime = currentTime
            }
            actualRefresh()
        }
    }

    private fun actualRefresh() {
        if (metadataStore.loggedIn) {
            exceptionRestConnector.refreshExceptions(this)
        } else {
            Toast.makeText(activity.applicationContext, R.string.login_first, Toast.LENGTH_SHORT).show()
            onExceptionRefreshFinished()
        }
    }

    override fun onExceptionRefreshFinished() {
        swipeRefreshLayout.isRefreshing = false
        refreshing = false
    }


    private fun initRecyclerView(inflater: LayoutInflater?, container: ViewGroup?): View {
        val fragmentView = inflater!!.inflate(R.layout.fragment_exception_instances, container, false)
        recyclerView = fragmentView.findViewById(R.id.exception_recycler_view) as RecyclerView
        recyclerView.layoutManager = LinearLayoutManager(activity)
        recyclerView.adapter = getAdapter(recyclerView)
        return fragmentView
    }

    private fun getAdapter(recyclerView: RecyclerView): ScaleInAnimationAdapter {
        val values = generateValues()
        adapter = ScaleInAnimationAdapter(ExceptionInstanceAdapter(values, recyclerView))
        adapter.notifyDataSetChanged()
        adapter.setDuration(200)
        adapter.setFirstOnly(true)
        return adapter
    }

    private fun generateValues(): List<Exception> {
        val allValues = exceptionInstanceStore.getExceptionList()
        if (friend != null) {
            val filteredValues = ArrayList<Exception>()
            for (exception in allValues) {
                if (exception.fromWho == friend!!.id || exception.toWho == friend!!.id) {
                    filteredValues.add(exception)
                }
            }
            return filteredValues
        } else {
            return allValues
        }
    }

    companion object {
        private var lastSyncTime: Long = 0
    }
}
